extensions/resource-deployment/src/helpers/cacheManager.ts

```diff
@@ -0,0 +1,83 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Deferred } from './promise';
+
+const enum Status {
+	notStarted,
+	inProgress,
+	done
+}
+
+interface State<T> {
+	entry?: T,
+	error?: Error,
+	status: Status,
+	id: number,
+	pendingOperation: Deferred<void>
+}
+
+/**
+ * An implementation of Cache Manager which ensures that only one call to populate cache miss is pending at a given time.
+ * All remaining calls for retrieval are awaited until the one in progress finishes and then all awaited calls are resolved with the value
+ * from the cache.
+ */
+export class CacheManager<K, T> {
```

Comment:
So why is this whole cache thing necessary? The way I see it we should just be fetching the valid options once upon opening the dialog and then using those to populate the dropdown. I don't really see the value in caching those options then - it just adds complexity for very little gain.

Reply:
it prevents multiple popups from being presented to the user for providing controller password when multiple calls to get the password value are received. The way the notebook wizard works when the summary page gets generated, we do get multiple calls for this password request. This cache manager serializes them and returns the same value that is fetched from the user once. Without it, a lot of code will need to be redone.

Reply:
> So why is this whole cache thing necessary? The way I see it we should just be fetching the valid options once upon opening the dialog and then using those to populate the dropdown. I don't really see the value in caching those options then - it just adds complexity for very little gain.

This would mean that if there are 5 controllers connected to by the user with none of them having a password saved, we would prompt the user for a password for each before opening the MIAA dialog. This is the behavior I am trying to avoid.

Reply:
Why would we need to fetch the password multiple times? The way I was thinking it'd work is that when the user selects the controller we get the password value then and validate it - if it's empty/wrong then we ask them to re-enter it at that point (and don't let them continue until they do and it passes validation).

Doing this shouldn't require any sort of cache/synchronizer since we only run that logic once when the data controller is picked.

Reply:
OK writing a handler for the dropdown picker to do the evaluation of subfields of the controller and saving those.

Reply:
Charles and I had an offline conversation, picking the sub-values on dropdown picker value change (values of password, username, the endpoint for the picked-up controller in our use case) can have its own user experience issues. The plan is to go with the current implementation and revisit this later.